### PR TITLE
port account signup to backend-v2

### PR DESCRIPTION
**Pull request by wginsberg**
part of https://github.com/Couchers-org/web-frontend/issues/227
related backend pr https://github.com/Couchers-org/backend-v2/pull/62

A few things depend on more backend work
* username verification happens too late in the signup flow, so using `"test"` will not work well when trying this out ;) - see https://github.com/Couchers-org/backend-v2/issues/64
* ~~community guidelines are hardcoded - see https://github.com/Couchers-org/backend-v2/issues/66~~
* no tracking of email verification state, so you have to remember to activate your account before you can `login` your user - see https://github.com/Couchers-org/backend-v2/pull/62#issuecomment-1312543104
* incidentally I had to change the `/contribute` page, but it depends on https://github.com/Couchers-org/backend-v2/issues/65 to be fully working

**Web frontend checklist**
- [x] Formatted my code with `make format`
- [x] There are no warnings from `make lint`
- [ ] There are no console warnings when running the app
- [ ] Added any new components to storybook
- [x] Added tests where relevant
- [x] All tests pass
- [x] Clicked around my changes running locally and it works
- [ ] Checked Desktop, Mobile and Tablet screen sizes

### CODE REVIEW

File: service/resources.ts [hunk 1]
```diff
@@ -1,14 +1,44 @@
...
+      icon: "shield.svg",
+    },
+    {
+      title: "Keep money out of it",
+      guideline:
+        "Keep things non-transactional. Don't exchange money while staying or hosting. Couchers.org is about learning, experiences, and getting to know people.",
+      icon: "handshake.svg",
+    },
+  ];
 }
```
Comment (darrenvong) [inline service/resources.ts:42]: I wonder if this is better suited to live in the frontend in one of the translation JSON file instead, where each version is linked to a different list of guidelines so we can use the version number retrieved from the backend to retrieve the right guidelines. After all, this feels like "static resource" to me where if we keep it at the frontend, we can also translate this to multiple languages which currently isn't possible since it comes from the backend.

**Review by darrenvong — COMMENTED** [review]
Another thing that's a bit confusing to me - if the user closes the page and comes back later, how does the sign up form know to take them back to the relevant step of the sign up flow/form? It's unclear where/how you are retrieving the saved state to get the user back onto the right form, which was done in some horrible `useEffect` before that you've removed in a previous PR but I can't see where the replacement for that is.

File: features/auth/signup/AccountForm.test.tsx [hunk 2]
```diff
@@ -73,12 +68,6 @@ describe("AccountForm", () => {
         ),
         "test"
       );
-      userEvent.type(
-        await screen.findByLabelText(
-          t("auth:account_form.password.field_label")
-        ),
-        "a very insecure password"
-      );
```
[the remark this answers is not in the document]
Comment (wginsberg) [reply features/auth/signup/AccountForm.test.tsx:81]: Yep, the password is set already in the basic form

File: features/auth/signup/Signup.test.tsx [hunk 3]
```diff
@@ -408,38 +328,23 @@ describe("Signup", () => {
...
-      needAccount: false,
-      needAcceptCommunityGuidelines: false,
-      needFeedback: false,
-      needVerifyEmail: false,
-      flowToken: "token",
-    };
-    window.localStorage.setItem("auth.flowState", JSON.stringify(state));
-    mockConsoleError();
-    await expect(async () => render(<View />, { wrapper })).rejects.toThrow();
-  });
```
[the remark this answers is not in the document]
Comment (wginsberg) [reply features/auth/signup/Signup.test.tsx:423]: Yep, once all of the required fields have been `PATCH`ed the response is just `{"user_created": true}`

File: features/auth/signup/Signup.test.tsx [hunk 4]
```diff
@@ -336,70 +288,38 @@ describe("Signup", () => {
...
-      needAccount: false,
-      needAcceptCommunityGuidelines: false,
-      needFeedback: false,
-      needVerifyEmail: true,
-      flowToken: "token",
-    };
-    window.localStorage.setItem("auth.flowState", JSON.stringify(state));
-    render(<View />, { wrapper });
-    expect(screen.getByText(t("auth:sign_up_completed_prompt"))).toBeVisible();
-  });
```
[the remark this answers is not in the document]
Comment (wginsberg) [reply features/auth/signup/Signup.test.tsx:392]: `"displays the feedback form when only feedback is pending"` is now covered by `"displays the feedback form when feedback is pending"`

`"displays the verify email message when email is pending"` can't be implemented at the moment - https://github.com/Couchers-org/backend-v2/pull/62#issuecomment-1312543104

File: service/resources.ts (hunk 1 above)
Comment (wginsberg) [reply service/resources.ts:42]: I'm not sure what the reason was for doing it this way in the first place
https://github.com/Couchers-org/couchers/issues/1682
I am indifferent on where it lives :P

**Comment by wginsberg** [comment]
> When I try to sign up from the beginning, I get the following error when the frontend POSTs to `/signup_flows`:
> 
> ```json
> {"error_messages":["Authentication credentials were not provided."],"status_code":401}
> ```
> 
> How did you get this to work? 🤔

Looks like there are some new changes to the API. Looks like we have email verification fields now too.

File: features/auth/useAuthStore.ts [hunk 5]
```diff
@@ -51,8 +59,10 @@ export default function useAuthStore() {
   );
   const [loading, setLoading] = useState(false);
   const [error, setError] = useState<string | null>(null);
-  const [flowState, setFlowState] =
-    usePersistedState<SignupFlowRes.AsObject | null>("auth.flowState", null);
+  const [flowState, setFlowState] = usePersistedState<SignupFlow | null>(
```
Comment (wginsberg) [inline features/auth/useAuthStore.ts:62]: @darrenvong this is where the flow state is read from localstorage. If I'm not wrong the `useEffect` you were referring to was only for the activation link

Comment (darrenvong) [reply features/auth/useAuthStore.ts:62]: But let's say the local storage is cleared, because the user went on the site for the first time, resuming from where they were before, and they have the flow token, how does that token helps to re-populate the state in memory and in local storage?

Unless I'm missing something, this seems to only work if they do the sign up all in one go and had the state stored in local storage already?

File: features/auth/signup/CommunityGuidelinesForm.tsx [hunk 6]
```diff
@@ -32,7 +32,7 @@ export default function CommunityGuidelinesForm() {
 
   const errorMessage = isHttpError(mutation.error)
     ? mutation.error?.error_messages[0]
-    : "";
+    : t("global:error.fatal_message");
```
Comment (darrenvong) [inline features/auth/signup/CommunityGuidelinesForm.tsx:35]: Getting a compile/type error here. I think it's just `error.fatal_message` or you'd need to wrap the namespace passed to `useTranslation` with a list.

**Comment by darrenvong** [comment]
> 



> > When I try to sign up from the beginning, I get the following error when the frontend POSTs to `/signup_flows`:
> > ```json
> > {"error_messages":["Authentication credentials were not provided."],"status_code":401}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > How did you get this to work? 🤔
> 
> Looks like there are some new changes to the API. Looks like we have email verification fields now too.

But how do we verify the email if we can't even send it to the API in the first place to start the sign up process? 😅

File: features/auth/signup/Signup.test.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (darrenvong) [reply features/auth/signup/Signup.test.tsx:392]: Does that mean we are a bit stuck on the frontend now then, since the user will need to activate their email separately before they finish the sign up process... but we have no way of telling them that right now unless they checked their email?

**Comment by wginsberg** [comment]
@darrenvong the auth error you were getting was because another branch got deployed to https://backend-v2-dev.couchers.dev/ - should be working again now

And I was wrong about new email validation fields, that was just something I saw in the swagger schema.

File: features/auth/useAuthStore.ts (hunk 5 above)
Comment (wginsberg) [reply features/auth/useAuthStore.ts:62]: If they're on a different device how would they have the token?

File: features/auth/signup/AccountForm.tsx [hunk 7]
```diff
@@ -79,26 +77,25 @@ export default function AccountForm() {
...
-        lng: location.lng,
-        radius: location.radius,
-      });
+      const state = await service.auth.signupFlowAccount(
+        authState.flowState!.flowToken,
+        {
+          username: lowercaseAndTrimField(username),
+          birthdate: birthdate.format().split("T")[0],
+          gender,
+          acceptedTOS: acceptTOS ? 1 : -1,
```
[the remark this answers is not in the document]
Comment (wginsberg) [reply features/auth/signup/AccountForm.tsx:92]: nope, I just thought it was nicer than making the field optional

File: features/auth/signup/Signup.test.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (wginsberg) [reply features/auth/signup/Signup.test.tsx:392]: Yes. I've added an issue for this on the backend-v2 repo.
https://github.com/Couchers-org/backend-v2/issues/68

File: features/auth/useAuthStore.ts (hunk 5 above)
Comment (darrenvong) [reply features/auth/useAuthStore.ts:62]: I believe they get it in the email in the old API's sign up flow, and by doing so, it also confirms their email since it's implied that if they have access to the token, they have access to the email

Comment (wginsberg) [reply features/auth/useAuthStore.ts:62]: Ah I see. With the new API the email is only sent after the signup flow is completed

File: service/resources.ts (hunk 1 above)
Comment (wginsberg) [reply service/resources.ts]: makes sense to me. changed in https://github.com/Couchers-org/web-frontend/pull/246/commits/0de5894fdac24e575f487d961b13f7d622218030

File: features/auth/useAuthStore.ts (hunk 5 above)
Comment (darrenvong) [reply features/auth/useAuthStore.ts:62]: It might be worth chatting with the backend team about this on another issue, as I think the whole point of having this flow token thing is so that user can resume sign up across devices or if they cleared their local storage/cookies. And I think that's meant to be a key feature of the sign up flow.
Otherwise, we may as well just store the WIP user ID we get from `POST /users` or something and keep patching the user object... with the caveat that if they clear their local storage/cookies, that user will effectively be "lost".

**Comment by darrenvong** [comment]
When you say the "happy path" works, does that include having to fill in the entire contributor form at the end of sign up? I got this error for trying to skip it:
```
{"errors":{"ideas":["This field may not be blank."],"features":["This field may not be blank."],"experience":["This field may not be blank."],"contribute":["This field may not be null."],"contribute_ways":["This list may not be empty."],"expertise":["This field may not be blank."]},"error_messages":["The data submitted was invalid"],"status_code":400}
```

And when you try fill in the form again (by refreshing), I get the user already exists error. In fact, even with the "happy path", I still get that error and not got an email... what else did you do that actually got it working? 🙈

File: features/auth/useAuthStore.ts [hunk 8]
```diff
@@ -118,21 +128,21 @@ export default function useAuthStore() {
...
           return;
         }
       },
-      async firstLogin(res: AuthRes.AsObject) {
-        setError(null);
-        setUserId(res.userId);
-        Sentry.setUser({ id: res.userId.toString() });
-        setJailed(res.jailed);
-        setAuthenticated(true);
-      },
```
Comment (darrenvong) [inline features/auth/useAuthStore.ts:135]: Why was this deleted?

Comment (wginsberg) [reply features/auth/useAuthStore.ts:135]: I guess because the account signup doesn't log the user in now there's no way to use this

**Review by darrenvong — APPROVED** [review]
Sorry for the delays in coming back to this!

It does look like the only thing not working in the happy path (skipping the contributor form) is a backend issue. Since we have a few of those at the moment anyway, maybe best to re-visit this area again once they're fixed like you said

File: features/auth/useAuthStore.ts (hunk 5 above)
Comment (wginsberg) [reply features/auth/useAuthStore.ts:62]: resolving, see conversation here -> https://github.com/Couchers-org/backend-v2/issues/60
